### DISCUSSION TITLE
What changed in the plugin protocol for 0.92.0?

### DISCUSSION THREAD

**Discussion by tesujimath** [nushell/nushell#12388] [General]
I just spent a bit of time bringing [nu_plugin_bash_env](https://github.com/tesujimath/nu_plugin_bash_env) up to date with the plugin protocol changes introduced in 0.91.0.

I had to abandon my API generator, as much of the plugin infrastructure has changed and become private. (I toyed with moving the API generator into the nu-plugin crate, but that was also not straightforward.)

So in the end I simply leaned on the description of the protocol for [Python plugins](https://www.nushell.sh/contributor-book/plugins.html#creating-a-plugin-in-python) and did it all by hand.  This was quite tricky TBH, but I made it work, so bash-env 0.6.0 supports Nushell 0.91.0. :tada: 

But now I see something changed again with the plugin protocol in Nushell 0.92.0, and I can't see what it was.  I think the description on that Python plugin page may now be outdated.

Is updating that one someone's TODO list?  It's really quite hard keeping a non-Rust plugin working!

I hope and expect the plugin protocol will stabilize soon!

**Comment by fdncred** [comment]
Have you read the release notes? https://www.nushell.sh/blog/2024-04-02-nushell_0_92_0.html You could also search for PRs by devyn and find a bunch of them related to plugins.

**Reply by tesujimath** [reply]
Oh!  Thanks, I hadn't seen that.  It's really useful. :flushed: 

I had in fact been wading through recent git commits related to plugins, but I overlooked that the PRs may have more commentary.

I'm actually in the process of adding some tracing to the plugin interface, which I am finding pretty much essential for going off-piste like this.  PR incoming soon.

Loving the progress in this area.  Thanks to all contributors! :sunglasses:

**Reply by fdncred** [reply]
There's also a contributors book that goes into detail about plugsins. There are two links https://www.nushell.sh/contributor-book/plugins.html and https://www.nushell.sh/contributor-book/plugin_protocol_reference.html

**Reply by tesujimath** [reply]
Yes I found those, thanks. And they're also helpful. 😊

**Reply by Abdillah** [reply]
The release note are super details. Unfortunately, it still hard to know what deserialization change affected on the new version. It would be great to note what the expected serialization especially `Value` (which only Int currently on the book) looks like..

**Reply by Abdillah** [reply]
How about a simple Rust deserialization tests?
It can mark and notice on some breaking change in the Plugin API, just like in https://github.com/nushell/nushell/pull/12326.

**Reply by fdncred** [reply]
@devyn has made many plugin changes recently and he's still working on some right now. i'm sure the plugin api will continue to change a little bit. more tests may be helpful. I'm sure devyn wouldn't mind other people jumping in with tests.

**Reply by devyn** [reply]
@Abdillah I absolutely have on my TODO list to document all of the `Value` types, so you will see that soon :slightly_smiling_face:

**Comment by tesujimath** [comment]
[This](https://github.com/nushell/nushell/pull/12403) is what I needed to add to Nu itself to work out how to get my plugin working again.

**Reply by fdncred** [reply]
I was thinking about your changes and our conversation and I think we do need some type of debugging ability. My fancy tui would be cool but I'm wondering if a more simplistic `trace!()` would be enough to get by for now? No env vars, just tracing the output.

**Reply by tesujimath** [reply]
The problem with that is the useful trace output gets swamped by everything else. There was a huge amount of log output when I enabled it.

@devyn 's idea of tee in a wrapper script could be the most effective way to see what's going on with an individual plugin. When I get a moment I'll try that out.

Of course, making this a general facility was what I was shooting for, but it would need to be separated out from other trace output.

**Reply by fdncred** [reply]
> There was a huge amount of log output when I enabled it.

Agreed, but we could make it a different log level. `trace!()` is huge, but some of the others are not.

I'm interested in the tee wrapper too. Would like to see it when you have time to work on it.

**Reply by tesujimath** [reply]
I finally got round to making the tracer plugin wrapper as per @devyn's idea.  It took me a while as trying to do this with `tee` and `bash` was an exercise in frustration, which I didn't even manage to solve with FIFOs or Bash coprocesses.

I solved it by writing a [wrapper in Rust](https://github.com/tesujimath/nu_plugin_tracer).  It works nicely.  Phew!

(There's a bit of an annoyance due to Nu enthusiastically canonicalizing paths, but I'll raise that as a separate issue.)

Shall I add this to the list of plugins on awesome Nu?  Even though it's not a plugin per se it would seem to be useful to list there.

**Reply by fdncred** [reply]
This is very cool. Nice work @tesujimath!

I'm wondering how it works on the latest main because `register` is being deprecated in favor of `plugin add`, which isn't just a rename but works completely differently now. plugin.nu is going away in favor of plugin.msgpackz and there are other plugin changes.

**Reply by tesujimath** [reply]
I had a quick look at `plugin add`, but I don't think its working sufficiently well for me to progress yet.  It doesn't seem to be able to let you add and use even one of the bundled plugins:
```
aya> cargo build --workspace
    Finished dev [unoptimized + debuginfo] target(s) in 0.38s
aya> cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
     Running `target/debug/nu`
aya> version
╭────────────────────┬─────────────────────────────────────────────────╮
│ version            │ 0.92.3                                          │
│ major              │ 0                                               │
│ minor              │ 92                                              │
│ patch              │ 3                                               │
│ branch             │ main                                            │
│ commit_hash        │ fbb4fad7ac8985751c12fd6a6382b867f7f6c095        │
│ build_os           │ linux-x86_64                                    │
│ build_target       │ x86_64-unknown-linux-gnu                        │
│ rust_version       │ rustc 1.77.2 (25ef9e3d8 2024-04-09)             │
│ cargo_version      │ cargo 1.77.2 (e52e36006 2024-03-26)             │
│ build_time         │ 1980-01-01 00:00:00 +00:00                      │
│ build_rust_channel │ debug                                           │
│ allocator          │ mimalloc                                        │
│ features           │ default, sqlite, system-clipboard, trash, which │
│ installed_plugins  │                                                 │
╰────────────────────┴─────────────────────────────────────────────────╯
aya> plugin list
╭────────────╮
│ empty list │
╰────────────╯
aya> plugin add target/debug/nu_plugin_formats
aya> plugin list
╭────────────╮
│ empty list │
╰────────────╯
```
Should list `nu_plugin_formats`, surely?

I'm happy to try again when this has stabilized sufficiently

**Reply by devyn** [reply]
Hi, sorry, `plugin add` is not exactly the same thing as `register`. We decided to split the command in to two parts because spawning plugin executables at parse time was undesirable. At the REPL, the equivalent is this:

```nushell
> plugin add target/debug/nu_plugin_formats
> plugin use formats
```

With `plugin use` now being the way to load something previously `add`ed. `plugin use` runs at parse-time like `register` did but it doesn't have to run a plugin executable, because it just relies on whatever's already been added to the plugin registry file. Of course, restarting nushell reloads everything as well, so you don't have to `plugin use` every time you want to use a plugin.

If you take a look at `help plugin` and the subcommands that should (hopefully) be descriptive enough, but I've also got PRs ready to update the docs on release, and the release notes will be pretty descriptive too.

**Reply by devyn** [reply]
Oh, and as a side note, I bet you can actually create a wrapper that abuses the `--shell` functionality. You can have nushell launch plugins with an auxilary program using

```nushell
plugin add --shell /path/to/program /path/to/nu_plugin_foo
```

**Reply by tesujimath** [reply]
Thanks @devyn I got this working now with the new plugin system.  I decided to use `--shell` as you suggest, which is much nicer than trying to infer the plugin name. :tada:
